doc/radosgw/s3-notification-compatibility.rst

```diff
@@ -79,31 +79,43 @@ However, the following fields may be sent empty, under the different deployment
-| ``s3:ObjectRestore:Post``                    | Not applicable to Ceph                                      |
-+----------------------------------------------+-----------------+-------------------------------------------+
-| ``s3:ObjectRestore:Complete``                | Not applicable to Ceph                                      |
-+----------------------------------------------+-----------------+-------------------------------------------+
-| ``s3:ReducedRedundancyLostObject``           | Not applicable to Ceph                                      |
-+----------------------------------------------+-----------------+-------------------------------------------+
++------------------------------------------------+-----------------+-------------------------------------------+
+| Event                                          | Notification    | PubSub                                    |
++================================================+=================+===========================================+
+| ``s3:ObjectCreated:*``                         | Supported                                                   |
++------------------------------------------------+-----------------+-------------------------------------------+
+| ``s3:ObjectCreated:Put``                       | Supported       | Supported at ``s3:ObjectCreated:*`` level |
++------------------------------------------------+-----------------+-------------------------------------------+
+| ``s3:ObjectCreated:Post``                      | Supported       | Not Supported                             |
++------------------------------------------------+-----------------+-------------------------------------------+
+| ``s3:ObjectCreated:Copy``                      | Supported       | Supported at ``s3:ObjectCreated:*`` level |
++------------------------------------------------+-----------------+-------------------------------------------+
+| ``s3:ObjectCreated:CompleteMultipartUpload``   | Supported       | Supported at ``s3:ObjectCreated:*`` level |
++------------------------------------------------+-----------------+-------------------------------------------+
+| ``s3:ObjectRemoved:*``                         | Supported       | Supported only the specific events below  |
++------------------------------------------------+-----------------+-------------------------------------------+
+| ``s3:ObjectRemoved:Delete``                    | Supported                                                   |
++------------------------------------------------+-----------------+-------------------------------------------+
+| ``s3:ObjectRemoved:DeleteMarkerCreated``       | Supported                                                   |
++------------------------------------------------+-----------------+-------------------------------------------+
+| ``s3:ObjectLifecycle:Expiration:Current``      | Supported, Ceph extension                                   |
```

Comment:
do we want to support `s3:ObjectLifecycle:*` ?
and/or `s3:ObjectLifecycle:Expiration:*` and `s3:ObjectLifecycle:Transition:*` ?
will it be common to ask for more than one type at once?

Reply:
yes, I've thought about this for a while, and think it makes sense

Reply:
> do we have an example of how to set lifecycle policy on a bucket, so we can test the new feature? I tried the following inside `src/test/rgw/bucket_notification/test_bn.py`:
> 
> ```python
>  response = client.put_bucket_lifecycle_configuration(Bucket=bucket_name, 
>              LifecycleConfiguration={'Rules': [
>                  {
>                      'Expiration': {'Date': datetime.datetime(2021, 12, 29), 'ExpiredObjectDeleteMarker': False},
>                      'Filter': {'Prefix': obj_prefix},
>                      'Status': 'Enabled',
>                  }
>              ]
>          }
>      )
> ```
> 
> and see:
> 
> ```
> s3:put_lifecycle read len=289 data=<LifecycleConfiguration xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Rule><Expiration><Date>2021-12-29T00:00:00Z</Date><ExpiredObjectDeleteMarker>false</ExpiredObjectDeleteMarker></Expiration><Filter><Prefix>ooo</Prefix></Filter><Status>Enabled</Status></Rule></LifecycleConfiguration>
> 
> s3:put_lifecycle Bad lifecycle configuration: LifecycleConfiguration: Rule: Expiration: bad Expiration section
> ```
> 
> in the rgw log

I tested with the following golang program:

package main

import (
	"fmt"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
	"github.com/aws/aws-sdk-go/service/sns"
)

func main() {

	fmt.Println("started")

	access_key := "0555b35654ad1656d804"
	secret_key := "h7GhxuBLTrlhVUyxSPUKUV8r/2EI4ngqJxD7iBdBYLhwluN30JaT3Q=="

	region := "us-east-1"
	host := "lemon:8000"
	force_pathstyle := true
	disable_ssl := true
	loglevel := aws.LogDebug

	config := &aws.Config{
		Region:   &region,
		Endpoint: &host,
		Credentials: credentials.NewStaticCredentials(
			access_key, secret_key, "" /* tokn */),
		S3ForcePathStyle: &force_pathstyle,
		DisableSSL:       &disable_ssl,
		LogLevel:         &loglevel,
	}

	// create an S3 handle
	session := session.New(config)
	conn := s3.New(session)

	// do it
	var params *s3.ListBucketsInput
	rsp, err := conn.ListBuckets(params)

	if err != nil {
		fmt.Println(err.Error())
	} else {
		fmt.Println(rsp)
	}

	topic_bucket := "bundeswehr"

	// create bucket
	cbp := &s3.CreateBucketInput{
		Bucket: aws.String(topic_bucket),
	}
	cbp_rsp, err := conn.CreateBucket(cbp)
	if err != nil {
		fmt.Println(err.Error())
	} else {
		fmt.Println(cbp_rsp)
	}

	// add kafka topic
	/*
		[&Attributes.entry.1.key=amqp-exchange&Attributes.entry.1.value=<exchange>]
		[&Attributes.entry.2.key=amqp-ack-level&Attributes.entry.2.value=none|broker|routable]
		[&Attributes.entry.3.key=verify-ssl&Attributes.entry.3.value=true|false]
		[&Attributes.entry.4.key=kafka-ack-level&Attributes.entry.4.value=none|broker]
		[&Attributes.entry.5.key=use-ssl&Attributes.entry.5.value=true|false]
		[&Attributes.entry.6.key=ca-location&Attributes.entry.6.value=<file path>]
		[&Attributes.entry.7.key=OpaqueData&Attributes.entry.7.value=<opaque data>]
		[&Attributes.entry.8.key=push-endpoint&Attributes.entry.8.value=<endpoint>]
		[&Attributes.entry.9.key=persistent&Attributes.entry.9.value=true|false]` */
	sconn := sns.New(session)
	ct_rsp, err := sconn.CreateTopic(&sns.CreateTopicInput{
		Name: aws.String(topic_bucket),
		Attributes: map[string]*string{
			"verify-ssl":      aws.String("false"),
			"use-ssl":         aws.String("false"),
			"kafka-ack-level": aws.String("none"),
			"push-endpoint":   aws.String("kafka://lemon:9092"),
			"persistent":      aws.String("false"), // for now!
		},
	})
	if err != nil {
		fmt.Println(err.Error())
	}
	fmt.Println(*ct_rsp.TopicArn)

	pbn_input := &s3.PutBucketNotificationConfigurationInput{
		Bucket: aws.String(topic_bucket),
		NotificationConfiguration: &s3.NotificationConfiguration{
			TopicConfigurations: []*s3.TopicConfiguration{
				{
					Id: aws.String(fmt.Sprintf("%s_%d", topic_bucket, 1)),
					Events: []*string{
						aws.String("s3:ObjectCreated:*"),
						aws.String("s3:ObjectRemoved:*"),
						aws.String("s3:ObjectLifecycle:*"), // RGW extension
					},
					TopicArn: aws.String(fmt.Sprintf("arn:aws:sns:default::%s", topic_bucket)),
				},
			},
		},
	}
	fmt.Println(pbn_input)

	pbnc_rsp, err := conn.PutBucketNotificationConfiguration(pbn_input)
	if err != nil {
		fmt.Println(err.Error())
	}
	fmt.Println(pbnc_rsp)

} /* main */

Reply:
but i dont see that you set the lifecycle configuration on the bucket

Reply:
oops, I missed one of your comments from 8/31 about eventName, I'm about to repush this

Reply:
oh, yeah, I did that in some python I think, just a sec

Reply:
sorry, I should have done more in the golang practice, just skipped to running the code, there's so many ways to set lifecycle

Reply:
#!/usr/bin/python

import boto3
import logging
import sys
import os
import time
import datetime
import pytz
import re
import pdb

rgw_host = os.environ['RGW_HOST']
rgw_port = int(os.environ['RGW_PORT'])
access_key = os.environ['RGW_ACCESS_KEY']
secret_key = os.environ['RGW_SECRET_KEY']

endpoint='http://%s:%d' % (rgw_host, rgw_port)

client = boto3.client('s3',
                      endpoint_url=endpoint,
                      aws_access_key_id=access_key,
                      aws_secret_access_key=secret_key)

nobjects = 100
bname = "bundeswehr"

bucket = client.create_bucket(Bucket=bname)

print(bname)
for o in range(0,nobjects):
        oname = "%s_o%d" % (bname, o)
        print(oname)
        oval = "value for me=%s" % oname
        client.put_object(Bucket=bname, Key=oname, Body=oval)

try:
    client.delete_bucket_lifecycle(Bucket=bname)
except:
    pass

rules=[{'ID': 'rule1', 'Expiration': {'Days': 1}, 'Prefix': '', 'Status':'Enabled'}]

lifecycle = {'Rules': rules}
res = client.put_bucket_lifecycle_configuration(
    Bucket=bname,
    LifecycleConfiguration=lifecycle)

Reply:
so I see, you're having issues with the Expiration form;  nothings changed there.
I'm pretty sure I've used this in the past:
```
rules=[{'ID': 'rule1', 'Expiration': { "Date": "20200101"}, 'Prefix': 'days0/',
        'Status':'Enabled'}]
lifecycle = {'Rules': rules}
res = client.put_bucket_lifecycle_configuration(
    Bucket=bucket_name,
    LifecycleConfiguration=lifecycle)
```